lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.memory.context;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ *  This class prevents contention in the memory tracking system by coarsening the granularity of memory tracking.
+ *  In this paradigm, the most small incremental increases at the byte granularity will not actually result
+ *  in a different coarse granularity value, so no reporting into memory tracking is necessary
+ */
+public class CoarseGrainLocalMemoryContext
+        implements LocalMemoryContext
+{
+    public static final long DEFAULT_GRANULARITY = 65536;
+
+    private final LocalMemoryContext delegate;
+    private final long granularity;
+    private final long mask;
+    private long currentBytes;
```

Comment:
```
    @GuardedBy("this")
```
is missing

Reply:
Added in https://github.com/trinodb/trino/pull/22881